### PR TITLE
Group upgrades komodo check pypi

### LINKED ISSUE #365: Group upgrades from komodo-check-pypi in levels
As a release manager, one would want to quickly have an overview of the potential packages to upgrade.
This is done via:
`komodo-check-pypi releases/matrices/2023.xx.xx.yml repository.yml --python-version 3.8`

The result is not structured yet, making it difficult to propose upgrades efficiently.
If we have major.minor.patch as tag, we would like to have groups of packages with patch, minor and major upgrades available.

Issue discussed during the Lessons Learned for 2023.05 release with the scout team.

As for now the result is difficulty to digest:

komodo-check-pypi releases/matrices/2023.05.rc0.yml repository.yml --python-version 3.8

Checking against python version: 3.8
alembic not at latest pypi version: 1.10.4, is at: 1.10.3
aniso8601 not at latest pypi version: 9.0.1, is at: 7.0.0
bokeh not at latest pypi version: 3.1.0, is at: 2.4.3
dash-bootstrap-components not at latest pypi version: 1.4.1, is at: 1.3.1
dask not at latest pypi version: 2023.4.0, is at: 2023.2.0
debugpy not at latest pypi version: 1.6.7, is at: 1.6.6
distributed not at latest pypi version: 2023.4.0, is at: 2023.2.0
fastapi not at latest pypi version: 0.95.1, is at: 0.92.0
fastjsonschema not at latest pypi version: 2.16.3, is at: 2.16.2
filelock not at latest pypi version: 3.12.0, is at: 3.9.0
flask not at latest pypi version: 2.3.1, is at: 2.2.2
fsspec not at latest pypi version: 2023.4.0, is at: 2023.1.0
geojson not at latest pypi version: 3.0.1, is at: 2.5.0
google-auth not at latest pypi version: 2.17.3, is at: 2.16.0
googleapis-common-protos not at latest pypi version: 1.59.0, is at: 1.58.0
gql not at latest pypi version: 3.4.0, is at: 2.0.0
graphene not at latest pypi version: 3.2.2, is at: 2.1.9
graphql-core not at latest pypi version: 3.2.3, is at: 2.3.2
graphql-relay not at latest pypi version: 3.2.0, is at: 2.0.1
greenlet not at latest pypi version: 2.0.2, is at: 1.1.3
grpcio not at latest pypi version: 1.54.0, is at: 1.51.1
hdf5plugin not at latest pypi version: 4.1.1, is at: 3.3.1
httpcore not at latest pypi version: 0.17.0, is at: 0.16.3
importlib-metadata not at latest pypi version: 6.6.0, is at: 4.12.0
importlib-resources not at latest pypi version: 5.12.0, is at: 5.10.2
ipykernel not at latest pypi version: 6.22.0, is at: 6.21.1
ipython not at latest pypi version: 8.12.0, is at: 8.10.0
ipywidgets not at latest pypi version: 8.0.6, is at: 8.0.4
jupyter-client not at latest pypi version: 8.2.0, is at: 7.3.5
jupyter-console not at latest pypi version: 6.6.3, is at: 6.5.0
jupyter-core not at latest pypi version: 5.3.0, is at: 4.12.0
jupyterlab-widgets not at latest pypi version: 3.0.7, is at: 3.0.5
markdown not at latest pypi version: 3.4.3, is at: 3.4.1
msal not at latest pypi version: 1.22.0, is at: 1.21.0
msgpack not at latest pypi version: 1.0.5, is at: 1.0.4
mypy-extensions not at latest pypi version: 1.0.0, is at: 0.4.3
natsort not at latest pypi version: 8.3.1, is at: 8.2.0
nbclient not at latest pypi version: 0.7.4, is at: 0.7.2
nbconvert not at latest pypi version: 7.3.1, is at: 7.2.9
nbformat not at latest pypi version: 5.8.0, is at: 5.7.3
notebook not at latest pypi version: 6.5.4, is at: 6.4.12
numpy not at latest pypi version: 1.24.3, is at: 1.24.2
opencensus not at latest pypi version: 0.11.2, is at: 0.11.1
opencensus-ext-azure not at latest pypi version: 1.1.9, is at: 1.1.8
openpyxl not at latest pypi version: 3.1.2, is at: 3.1.0
openvds not at latest pypi version: 3.2.3, is at: 3.2.1
opm not at latest pypi version: 2022.10, is at: 2022.4
orjson not at latest pypi version: 3.8.11, is at: 3.8.10
ortools not at latest pypi version: 9.6.2534, is at: 9.4.1874
packaging not at latest pypi version: 23.1, is at: 23.0
pandas not at latest pypi version: 2.0.1, is at: 1.5.3
partd not at latest pypi version: 1.4.0, is at: 1.3.0
prefect not at latest pypi version: 2.10.6, is at: 0.15.13
prompt-toolkit not at latest pypi version: 3.0.38, is at: 3.0.36
protobuf not at latest pypi version: 4.22.3, is at: 3.19.5
psutil not at latest pypi version: 5.9.5, is at: 5.9.4
pyasn1 not at latest pypi version: 0.5.0, is at: 0.4.8
pyasn1-modules not at latest pypi version: 0.3.0, is at: 0.2.8
pybind11 not at latest pypi version: 2.10.4, is at: 2.10.3
pydantic not at latest pypi version: 1.10.7, is at: 1.10.4
pygments not at latest pypi version: 2.15.1, is at: 2.14.0
pyqt5-sip not at latest pypi version: 12.12.1, is at: 12.11.1
python-box not at latest pypi version: 7.0.1, is at: 6.0.2
python-multipart not at latest pypi version: 0.0.6, is at: 0.0.5
python-slugify not at latest pypi version: 8.0.1, is at: 6.1.2
pytz not at latest pypi version: 2023.3, is at: 2022.7.1
pyzmq not at latest pypi version: 25.0.2, is at: 24.0.1
qtconsole not at latest pypi version: 5.4.2, is at: 5.4.0
qtpy not at latest pypi version: 2.3.1, is at: 2.3.0
requests not at latest pypi version: 2.29.0, is at: 2.28.2
rips not at latest pypi version: 2023.3.0.1, is at: 2023.1.0.1
rx not at latest pypi version: 3.2.0, is at: 1.6.1
scikit-build not at latest pypi version: 0.17.2, is at: 0.16.6
scikit-learn not at latest pypi version: 1.2.2, is at: 1.2.1
scipy not at latest pypi version: 1.10.1, is at: 1.9.1
send2trash not at latest pypi version: 1.8.2, is at: 1.8.0
setuptools not at latest pypi version: 67.7.2, is at: 66.1.0
singledispatch not at latest pypi version: 4.0.0, is at: 3.7.0
soupsieve not at latest pypi version: 2.4.1, is at: 2.3.2.post1
sphinx not at latest pypi version: 6.2.1, is at: 5.2.1
sqlalchemy not at latest pypi version: 2.0.11, is at: 1.4.41
starlette not at latest pypi version: 0.26.1, is at: 0.25.0
tornado not at latest pypi version: 6.3.1, is at: 6.2
treelib not at latest pypi version: 1.6.4, is at: 1.6.1
urllib3 not at latest pypi version: 2.0.0, is at: 1.26.14
uvicorn not at latest pypi version: 0.22.0, is at: 0.20.0
websockets not at latest pypi version: 11.0.2, is at: 10.4
webviz-subsurface not at latest pypi version: 0.2.18, is at: 0.2.19rc0
werkzeug not at latest pypi version: 2.3.1, is at: 2.2.3
wheel not at latest pypi version: 0.40.0, is at: 0.38.4
widgetsnbextension not at latest pypi version: 4.0.7, is at: 4.0.5
wslink not at latest pypi version: 1.10.1, is at: 1.10.0
xtgeo not at latest pypi version: 3.0.1, is at: 3.0.0
yarl not at latest pypi version: 1.9.2, is at: 1.8.2
zict not at latest pypi version: 3.0.0, is at: 2.2.0

### PR DESCRIPTION
**Pull request by jonathan-eq**
Resolves #365 
Should not depend on but,
Depends on #388

### CODE REVIEW

File: komodo/check_up_to_date_pypi.py [hunk 1]
```diff
@@ -96,7 +96,11 @@ def get_upgrade_proposals_from_pypi(
 ) -> dict:
     pypi_packages = get_pypi_packages(releases, repository)
     pypi_responses = get_pypi_info(pypi_packages)
+    pypi_responses = get_pypi_info(pypi_packages)
 
+    upgrade_proposals_from_pypi = {}
+    for package_name, _response in pypi_responses:
+        komodo_version = get_version.parse(strip_version(releases[package_name]))
```
Comment (berland) [inline komodo/check_up_to_date_pypi.py:103]: Duplicated code?

File: komodo/check_up_to_date_pypi.py [hunk 2]
```diff
@@ -158,19 +162,47 @@ def run_check_up_to_date(
                 "Writing upgrade proposals from pypi, "
                 "assuming nothing has changed with dependencies..."
             )
-            with open(propose_upgrade, mode="w", encoding="utf-8") as fout:
+            with open(propose_upgrade, "w") as fout:
```
Comment (berland) [inline komodo/check_up_to_date_pypi.py:165]: keep encoding.

File: komodo/check_up_to_date_pypi.py [hunk 3]
```diff
@@ -158,19 +162,47 @@ def run_check_up_to_date(
...
+                patch_upgrades.append(
+                    f"{name} not at latest pypi version: {pypi_latest}, "
+                    f"is at: {current_version}"
+                )
+            else:
+                other_upgrades.append(
+                    f"{name} not at latest pypi version: {pypi_latest}, "
+                    f"is at: {current_version}"
+                )
+        sys.exit(
```
Comment (berland) [inline komodo/check_up_to_date_pypi.py:193]: Nonzero argument to `sys.exi`t usually means something bad happened. I would prefer `print` statements, and then followed by `sys.exit(someinteger)`. Then we can discuss whether upgrades being present is something that should give a nonzero exit value. Do we want some script (e.g. a github workflow) to stop if there are upgrades present? If so, that would argue for a nonzero exit value.

Comment (berland) [reply komodo/check_up_to_date_pypi.py:193]: (changing from sys.exit to print is really unrelated to the issue you are solving, so you could prepend your PR with a commit that first changes from sys.exit to print)

File: komodo/check_up_to_date_pypi.py [hunk 4]
```diff
@@ -232,20 +264,12 @@ def main():
 
     print(f"Checking against python version: {args.python_version}")
 
-    validate_release_file(args.release_file)
-    validate_repository_file(args.repository_file)
+    ReleaseFile()(args.release_file)
```
Comment (berland) [inline komodo/check_up_to_date_pypi.py:267]: previous code was more readable! :)

File: komodo/check_up_to_date_pypi.py (hunk 1 above)
Comment (jonathan-eq) [reply komodo/check_up_to_date_pypi.py:103]: there was a merge conflict, so it kept both changes...

Comment (jonathan-eq) [reply komodo/check_up_to_date_pypi.py]: I think I got all of the duplicates

**Review by berland — APPROVED** [review]
Good!

**Pull request merged**